docs/ispc.rst

```diff
@@ -632,6 +632,20 @@ Updating ISPC Programs For Changes In ISPC 1.25.0
 
 ``avx512knl-x16`` target is deprecated and will be removed in future releases.
 
+ISPC language was extended to support ``__attribute__(())`` syntax for variable
+and function declarations. Two attributes are supported now: ``noescape`` and
+``address_space(N)``. The macro ``ISPC_ATTRIBUTE_SUPPORTED`` is defined if the
+ISPC compiler supports these attributes.
+
+``__attribute__((noescape))`` can be placed on a function parameter of a
+uniform pointer type. It informs the compiler that the pointer cannot escape,
+i.e., no reference to the object the pointer points to, derived from the
+parameter value, will survive after the function returns.
+
+``__attribute__((address_space(N)))`` can be applied to a variable or parameter
```

Comment:
Can these attributes be applied to function pointers?

I tried this and it doesn't compile:
```
#define ADDR_SPACE(N) __attribute__((address_space(N)))

void foo(int a) {
    print("foo: %d\n", a);
}

void bar(int a) {
    print("bar: %d\n", a);
}

typedef void (ADDR_SPACE(1) *func_ptr_t)(int);

void call_function(func_ptr_t func, int value) {
    func(value);
}

void test() {
    func_ptr_t fp1 = (func_ptr_t)foo;
    func_ptr_t fp2 = (func_ptr_t)bar;

    call_function(fp1, 42);
    call_function(fp2, 84);

    }
```

Reply:
This code doesn't seem to work in clang either https://godbolt.org/z/bYvE5vvPb

Reply:
Here is fixed example: https://godbolt.org/z/5TqEjn9q3

Reply:
> Should the attribute be always before type? 

No, it should not. Moreover, the meaning can be different as it is for `address_space`. Consider the following example: https://godbolt.org/z/YYqeqo3eP.

Reply:
So I see that the following code is valid in C++ but ISPC produces an error
```
void foo(int a) {
    print("foo: %d\n", a);
}
typedef void (*func_ptr_t)(int);
ADDR_SPACE(1) func_ptr_t f1 = foo;

 Error: Can't convert between pointer types with different address spaces " void(varying int32 a) * uniform" and " void(varying int32 __anon_parameter_0) addrspace(1) * varying" for initializer.
ADDR_SPACE(1) func_ptr_t f1 = foo;
```
I think the error reporting for attributes with function pointers may be improved. Or we can prohibit the usage of it with function pointers and explicitly say about it in docs.

Reply:
But the error is valid because foo is not placed in `address_space = 1`. Consider that this is absolutely valid:
```ispc
ADDR_SPACE(0) func_ptr_t f1 = foo;
```

Reply:
Ah, got it. OK!